comm/src/main/scala/coop/rchain/comm/transport/StreamHandler.scala

```diff
@@ -69,33 +67,43 @@ object StreamHandler {
     } yield Streamed(fos = fos, path = file)
 
   private def collect(
+      networkId: String,
       init: Streamed,
       stream: Observable[Chunk],
       circuitBreaker: CircuitBreaker
   ): EitherT[Task, Throwable, Streamed] = {
 
-    def collectStream = stream.foldWhileLeftL(init) {
-      case (stmd, Chunk(Chunk.Content.Header(ChunkHeader(sender, typeId, compressed, cl)))) =>
-        Left(
-          stmd.copy(
-            sender = sender.map(ProtocolHelper.toPeerNode(_)),
-            typeId = Some(typeId),
-            compressed = compressed,
-            contentLength = Some(cl)
-          )
-        )
-      case (stmd, Chunk(Chunk.Content.Data(ChunkData(newData)))) =>
-        val array = newData.toByteArray
-        stmd.fos.write(array)
-        stmd.fos.flush()
-        val readSoFar = stmd.readSoFar + array.length
-        if (circuitBreaker(readSoFar))
-          Right(stmd.copy(circuitBroken = true))
-        else
-          Left(stmd.copy(readSoFar = readSoFar))
-    }
+    def collectStream: Task[Streamed] =
+      stream.foldWhileLeftL(init) {
+        case (
+            stmd,
+            Chunk(Chunk.Content.Header(ChunkHeader(sender, typeId, compressed, cl, nid)))
+            ) =>
+          if (nid == networkId) {
+            Left(
+              stmd.copy(
+                sender = sender.map(ProtocolHelper.toPeerNode),
+                typeId = Some(typeId),
+                compressed = compressed,
+                contentLength = Some(cl)
+              )
+            )
+          } else
+            Right(stmd.copy(wrongNetwork = true))
+        case (stmd, Chunk(Chunk.Content.Data(ChunkData(newData)))) =>
+          val array = newData.toByteArray
+          stmd.fos.write(array)
+          stmd.fos.flush()
+          val readSoFar = stmd.readSoFar + array.length
+          if (circuitBreaker(readSoFar))
+            Right(stmd.copy(circuitBroken = true))
+          else
+            Left(stmd.copy(readSoFar = readSoFar))
+      }
 
     EitherT(collectStream.attempt.map {
+      case Right(stmd) if stmd.wrongNetwork =>
+        new RuntimeException("Wrong network id").asLeft
```

Comment:
Are you sure we want to throw an RuntimeException here? It would be a common thing once we start running multiple networks

Reply:
Actually, this exception won't get thrown. The private function `collect` is used only by `StreamHandler.handleStream` which returns a `Task[Either[Throwable, StreamMessage]]`. This is used then as
```scala
(StreamHandler.handleStream(networkId, tempFolder, observable, neverBreak) >>= {
  case Left(ex)   => logger.error(s"Could not receive stream! Details: ${ex.getMessage}", ex)
  case Right(msg) => Task.delay(bufferBlobMessage.pushNext(msg)).as(())
}).as(ChunkResponse())
```

On the other hand, once all nodes are updated to this version we should never see streams coming from other networks because they will get filtered at the node discovery layer. However, I implemented also the network ID check at the RChain protocol layer to prevent traffic from older nodes.